File: SubArraySunEqualsK.java
import java.util.*;

//Time Complexity : o(n) - total no of elements
//Space Complexity : o(n)
//Did this code successfully run on Leetcode : yes
public class SubArraySunEqualsK {

	public int subarraySum(int[] nums, int k) {

		if (nums == null || nums.length == 0)
			return 0;
		int count = 0, rSum = 0;
		// key - running sum, val - count
		HashMap<Integer, Integer> map = new HashMap<>();
		map.put(0, 1);
		for (int i = 0; i < nums.length; i++) {
			rSum += nums[i];
			// check map contains the complement
			int comp = rSum - k;
			if (map.containsKey(comp)) {
				// add no of counts to result
				count += map.get(comp);
			}
			// add or increase count for particular running sum for next iteration
			map.put(rSum, map.getOrDefault(rSum, 0) + 1);
		}
		return count;
	}

}

File: Subarray sun equals k.py
# Time Complexity = O(n) 
# Space Complexity = O(n) 
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

import collections
class Solution:
    def subarraySum(self, nums, k: int) -> int:
        
        # use counter to keep track of sums
        s, t, c = 0, 0, collections.Counter()
        for n in nums:
            c[s] += 1  # increment if s adds up to the key value
            s += n     # iterate through nums by adding elements 
            t += c[s - k]  # increament t(possible subarrays) if sum - k iteratively happens
        return t